From c4836c9b5996ead8422cc11395da1d29efd06392 Mon Sep 17 00:00:00 2001
From: Salvo Giangreco <giangrecosalvo9@gmail.com>
Date: Sat, 29 Aug 2026 17:30:34 +0200
Subject: [PATCH] Introduce CscFeatureHooks

UN1CA: Adapted for Android 13 (One UI 5.1).
SemCscFeature is in smali_classes5 on Android 13 instead of smali_classes6.

Co-authored-by: Android-Artisan <romartisan2025@gmail.com>
---
 .../android/os/SemSystemProperties.smali      |  10 +
 .../android/feature/SemCscFeature.smali       |  87 +++-
 .../io/mesalabs/unica/CscFeatureHooks.smali   | 381 ++++++++++++++++++
 3 files changed, 468 insertions(+), 10 deletions(-)
 create mode 100644 smali_classes6/io/mesalabs/unica/CscFeatureHooks.smali

diff --git a/smali_classes3/android/os/SemSystemProperties.smali b/smali_classes3/android/os/SemSystemProperties.smali
index 3770714..9329be9 100644
--- a/smali_classes3/android/os/SemSystemProperties.smali
+++ b/smali_classes3/android/os/SemSystemProperties.smali
@@ -45,6 +45,15 @@
 .method public static whitelist getCountryCode()Ljava/lang/String;
     .locals 2
 
+    invoke-static {}, Lio/mesalabs/unica/CscFeatureHooks;->onGetCountryCode()Ljava/lang/String;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_0
+
+    goto :goto_0
+
+    :cond_0
     const-string/jumbo v0, "ro.csc.country_code"
 
     const-string v1, ""
@@ -53,6 +62,7 @@
 
     move-result-object v0
 
+    :goto_0
     return-object v0
 .end method
 
diff --git a/smali_classes5/com/samsung/android/feature/SemCscFeature.smali b/smali_classes5/com/samsung/android/feature/SemCscFeature.smali
index f5ea00d..08d7420 100644
--- a/smali_classes5/com/samsung/android/feature/SemCscFeature.smali
+++ b/smali_classes5/com/samsung/android/feature/SemCscFeature.smali
@@ -2774,8 +2774,21 @@
 .end method
 
 .method public whitelist getBoolean(Ljava/lang/String;)Z
-    .locals 2
+    .locals 4
+
+    invoke-static {p1}, Lio/mesalabs/unica/CscFeatureHooks;->onGetBoolean(Ljava/lang/String;)Ljava/lang/Boolean;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_0
+
+    invoke-virtual {v0}, Ljava/lang/Boolean;->booleanValue()Z
+
+    move-result p0
 
+    return p0
+
+    :cond_0
     const/4 v0, 0x0
 
     :try_start_0
@@ -2807,8 +2820,21 @@
 .end method
 
 .method public whitelist getBoolean(Ljava/lang/String;Z)Z
-    .locals 2
+    .locals 3
+
+    invoke-static {p1}, Lio/mesalabs/unica/CscFeatureHooks;->onGetBoolean(Ljava/lang/String;)Ljava/lang/Boolean;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_0
+
+    invoke-virtual {v0}, Ljava/lang/Boolean;->booleanValue()Z
+
+    move-result p0
 
+    return p0
+
+    :cond_0
     :try_start_0
     iget-object v0, p0, Lcom/samsung/android/feature/SemCscFeature;->mFeatureList:Ljava/util/Hashtable;
 
@@ -2858,8 +2884,21 @@
 .end method
 
 .method public whitelist getInt(Ljava/lang/String;)I
-    .locals 2
+    .locals 4
 
+    invoke-static {p1}, Lio/mesalabs/unica/CscFeatureHooks;->onGetInt(Ljava/lang/String;)Ljava/lang/Integer;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_0
+
+    invoke-virtual {v0}, Ljava/lang/Integer;->intValue()I
+
+    move-result p0
+
+    return p0
+
+    :cond_0
     const/4 v0, -0x1
 
     :try_start_0
@@ -2891,8 +2930,21 @@
 .end method
 
 .method public whitelist getInt(Ljava/lang/String;I)I
-    .locals 2
+    .locals 3
+
+    invoke-static {p1}, Lio/mesalabs/unica/CscFeatureHooks;->onGetInt(Ljava/lang/String;)Ljava/lang/Integer;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_0
+
+    invoke-virtual {v0}, Ljava/lang/Integer;->intValue()I
+
+    move-result p0
+
+    return p0
 
+    :cond_0
     :try_start_0
     iget-object v0, p0, Lcom/samsung/android/feature/SemCscFeature;->mFeatureList:Ljava/util/Hashtable;
 
@@ -2982,7 +3034,13 @@
 .end method
 
 .method public whitelist getString(Ljava/lang/String;)Ljava/lang/String;
-    .locals 2
+    .locals 4
+
+    invoke-static {p1}, Lio/mesalabs/unica/CscFeatureHooks;->onGetString(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object v0
+
+    if-nez v0, :cond_1
 
     const-string v0, ""
 
@@ -2997,11 +3055,11 @@
     :try_end_0
     .catch Ljava/lang/Exception; {:try_start_0 .. :try_end_0} :catch_0
 
-    if-eqz v1, :cond_0
+    if-eqz v1, :cond_1
 
     return-object v1
 
-    :cond_0
+    :cond_1
     return-object v0
 
     :catch_0
@@ -3011,8 +3069,17 @@
 .end method
 
 .method public whitelist getString(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
-    .locals 1
+    .locals 3
 
+    invoke-static {p1}, Lio/mesalabs/unica/CscFeatureHooks;->onGetString(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_0
+
+    return-object v0
+
+    :cond_0
     :try_start_0
     iget-object v0, p0, Lcom/samsung/android/feature/SemCscFeature;->mFeatureList:Ljava/util/Hashtable;
 
@@ -3024,11 +3091,11 @@
     :try_end_0
     .catch Ljava/lang/Exception; {:try_start_0 .. :try_end_0} :catch_0
 
-    if-eqz v0, :cond_0
+    if-eqz v0, :cond_1
 
     return-object v0
 
-    :cond_0
+    :cond_1
     return-object p2
 
     :catch_0
diff --git a/smali_classes6/io/mesalabs/unica/CscFeatureHooks.smali b/smali_classes6/io/mesalabs/unica/CscFeatureHooks.smali
new file mode 100644
index 0000000..eb82116
--- /dev/null
+++ b/smali_classes6/io/mesalabs/unica/CscFeatureHooks.smali
@@ -0,0 +1,381 @@
+@@ -0,0 +1,380 @@
+.class public final Lio/mesalabs/unica/CscFeatureHooks;
+.super Ljava/lang/Object;
+.source "CscFeatureHooks.java"
+
+
+# static fields
+.field private static final blacklist BOOLEAN_FEATURES:Ljava/util/HashMap;
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "Ljava/util/HashMap<",
+            "Ljava/lang/String;",
+            "Ljava/lang/Boolean;",
+            ">;"
+        }
+    .end annotation
+.end field
+
+.field private static final blacklist CAMERA_PACKAGE_NAME:Ljava/lang/String; = "com.sec.android.app.camera"
+
+.field private static final blacklist DEBUG:Z = false
+
+.field private static final blacklist INTEGER_FEATURES:Ljava/util/HashMap;
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "Ljava/util/HashMap<",
+            "Ljava/lang/String;",
+            "Ljava/lang/Integer;",
+            ">;"
+        }
+    .end annotation
+.end field
+
+.field private static final blacklist STRING_FEATURES:Ljava/util/HashMap;
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "Ljava/util/HashMap<",
+            "Ljava/lang/String;",
+            "Ljava/lang/String;",
+            ">;"
+        }
+    .end annotation
+.end field
+
+.field private static final blacklist TAG:Ljava/lang/String; = "CscFeatureHooks"
+
+
+# direct methods
+.method static constructor blacklist <clinit>()V
+    .locals 5
+
+    new-instance v0, Ljava/util/HashMap;
+
+    invoke-direct {v0}, Ljava/util/HashMap;-><init>()V
+
+    sput-object v0, Lio/mesalabs/unica/CscFeatureHooks;->BOOLEAN_FEATURES:Ljava/util/HashMap;
+
+    new-instance v1, Ljava/util/HashMap;
+
+    invoke-direct {v1}, Ljava/util/HashMap;-><init>()V
+
+    sput-object v1, Lio/mesalabs/unica/CscFeatureHooks;->STRING_FEATURES:Ljava/util/HashMap;
+
+    new-instance v2, Ljava/util/HashMap;
+
+    invoke-direct {v2}, Ljava/util/HashMap;-><init>()V
+
+    sput-object v2, Lio/mesalabs/unica/CscFeatureHooks;->INTEGER_FEATURES:Ljava/util/HashMap;
+
+    const-string v3, "CscFeature_Common_SupportPrivateMode"
+
+    sget-object v4, Ljava/lang/Boolean;->TRUE:Ljava/lang/Boolean;
+
+    invoke-virtual {v0, v3, v4}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v3, "CscFeature_Setting_EnableHwVersionDisplay"
+
+    sget-object v4, Ljava/lang/Boolean;->TRUE:Ljava/lang/Boolean;
+
+    invoke-virtual {v0, v3, v4}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v3, "CscFeature_Wifi_SupportBlockSkipForSetupWizard"
+
+    sget-object v4, Ljava/lang/Boolean;->FALSE:Ljava/lang/Boolean;
+
+    invoke-virtual {v0, v3, v4}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v3, "CscFeature_Wifi_SupportNetworkConnectionsRequired"
+
+    sget-object v4, Ljava/lang/Boolean;->FALSE:Ljava/lang/Boolean;
+
+    invoke-virtual {v0, v3, v4}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v0, "CscFeature_Common_ConfigSvcProviderForUnknownNumber"
+
+    const-string v3, "whitepages,whitepages,off"
+
+    invoke-virtual {v1, v0, v3}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v0, "CscFeature_Common_ConfigYuva"
+
+    const-string v3, ""
+
+    invoke-virtual {v1, v0, v3}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v0, "CscFeature_VoiceCall_ConfigRecording"
+
+    const-string v4, "RecordingAllowed"
+
+    invoke-virtual {v1, v0, v4}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v0, "CscFeature_SmartManager_ConfigSubFeatures"
+
+    invoke-virtual {v1, v0, v3}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const/4 v0, 0x0
+
+    invoke-static {v0}, Ljava/lang/Integer;->valueOf(I)Ljava/lang/Integer;
+
+    move-result-object v0
+
+    const-string v1, "CscFeature_Setting_ConfigTypeHelp"
+
+    invoke-virtual {v2, v1, v0}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    return-void
+.end method
+
+.method private constructor blacklist <init>()V
+    .locals 0
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    return-void
+.end method
+
+.method private static blacklist dlog(Ljava/lang/String;)V
+    .locals 0
+
+    return-void
+.end method
+
+.method public static blacklist onGetBoolean(Ljava/lang/String;)Ljava/lang/Boolean;
+    .locals 3
+
+    sget-object v0, Lio/mesalabs/unica/CscFeatureHooks;->BOOLEAN_FEATURES:Ljava/util/HashMap;
+
+    invoke-virtual {v0, p0}, Ljava/util/HashMap;->containsKey(Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-eqz v1, :cond_0
+
+    invoke-virtual {v0, p0}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v0
+
+    check-cast v0, Ljava/lang/Boolean;
+
+    goto :goto_0
+
+    :cond_0
+    const/4 v0, 0x0
+
+    :goto_0
+    if-eqz v0, :cond_1
+
+    new-instance v1, Ljava/lang/StringBuilder;
+
+    const-string v2, "Spoofing \""
+
+    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v1, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\" to \""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0, v0}, Ljava/lang/StringBuilder;->append(Ljava/lang/Object;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0}, Lio/mesalabs/unica/CscFeatureHooks;->dlog(Ljava/lang/String;)V
+
+    :cond_1
+    return-object v0
+.end method
+
+.method public static blacklist onGetCountryCode()Ljava/lang/String;
+    .locals 3
+
+    invoke-static {}, Landroid/app/ActivityThread;->currentPackageName()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-static {v0}, Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_0
+
+    const-string v1, "com.sec.android.app.camera"
+
+    invoke-virtual {v0, v1}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_0
+
+    const-string v0, "USA"
+
+    goto :goto_0
+
+    :cond_0
+    const/4 v0, 0x0
+
+    :goto_0
+    if-eqz v0, :cond_1
+
+    new-instance v1, Ljava/lang/StringBuilder;
+
+    const-string v2, "Spoofing SemSystemProperties.getCountryCode() to \""
+
+    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v1, v0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object v1
+
+    const-string v2, "\""
+
+    invoke-virtual {v1, v2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object v1
+
+    invoke-virtual {v1}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object v1
+
+    invoke-static {v1}, Lio/mesalabs/unica/CscFeatureHooks;->dlog(Ljava/lang/String;)V
+
+    :cond_1
+    return-object v0
+.end method
+
+.method public static blacklist onGetInt(Ljava/lang/String;)Ljava/lang/Integer;
+    .locals 3
+
+    sget-object v0, Lio/mesalabs/unica/CscFeatureHooks;->INTEGER_FEATURES:Ljava/util/HashMap;
+
+    invoke-virtual {v0, p0}, Ljava/util/HashMap;->containsKey(Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-eqz v1, :cond_0
+
+    invoke-virtual {v0, p0}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v0
+
+    check-cast v0, Ljava/lang/Integer;
+
+    goto :goto_0
+
+    :cond_0
+    const/4 v0, 0x0
+
+    :goto_0
+    if-eqz v0, :cond_1
+
+    new-instance v1, Ljava/lang/StringBuilder;
+
+    const-string v2, "Spoofing \""
+
+    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v1, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\" to \""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0, v0}, Ljava/lang/StringBuilder;->append(Ljava/lang/Object;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0}, Lio/mesalabs/unica/CscFeatureHooks;->dlog(Ljava/lang/String;)V
+
+    :cond_1
+    return-object v0
+.end method
+
+.method public static blacklist onGetString(Ljava/lang/String;)Ljava/lang/String;
+    .locals 3
+
+    sget-object v0, Lio/mesalabs/unica/CscFeatureHooks;->STRING_FEATURES:Ljava/util/HashMap;
+
+    invoke-virtual {v0, p0}, Ljava/util/HashMap;->containsKey(Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-eqz v1, :cond_0
+
+    invoke-virtual {v0, p0}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v0
+
+    check-cast v0, Ljava/lang/String;
+
+    goto :goto_0
+
+    :cond_0
+    const/4 v0, 0x0
+
+    :goto_0
+    if-eqz v0, :cond_1
+
+    new-instance v1, Ljava/lang/StringBuilder;
+
+    const-string v2, "Spoofing \""
+
+    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v1, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\" to \""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0, v0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0}, Lio/mesalabs/unica/CscFeatureHooks;->dlog(Ljava/lang/String;)V
+
+    :cond_1
+    return-object v0
+.end method
-- 
2.53.0

